javalib/src/main/scala/java/util/StringJoiner.scala

```diff
@@ -1,76 +1,115 @@
 // Ported from Scala.js commit: 57d71da dated: 2023-05-31
-// Two methods are different, see below.
+// This file has many differences, using CharSequence instead of String,
+// and different implementations of add, merge, toString and length.
 
 package java.util
 
+import java.lang.StringBuilder
+
 @inline
 final class StringJoiner private (
-    delimiter: String,
-    prefix: String,
-    suffix: String
+    delimiter: CharSequence,
+    prefix: CharSequence,
+    suffix: CharSequence
 ) extends AnyRef {
 
-  /** The custom value to return if empty, set by `setEmptyValue` (nullable).
-   *
-   *  If `null`, defaults to `prefix + suffix`.
-   */
+  /** None of the arguments should be null. */
+  Objects.requireNonNull(delimiter, "The delimiter must not be null")
+  Objects.requireNonNull(prefix, "The prefix must not be null")
+  Objects.requireNonNull(suffix, "The suffix must not be null")
+
+  /** We need the fields to be immutable. */
+  private val delimStr: String = delimiter.toString()
+  private val prefixStr: String = prefix.toString()
+  private val suffixStr: String = suffix.toString()
+
+  /** The custom value to return if empty, set by `setEmptyValue` (nullable). */
   private var emptyValue: String = null
 
-  /** The current value, excluding prefix and suffix. */
-  private var value: String = ""
+  /** A list that holds (in reverse) strings that have been added so far. */
+  private var contents: scala.collection.immutable.List[CharSequence] = Nil
 
   /** Whether the string joiner is currently empty. */
   private var isEmpty: Boolean = true
 
-  def this(delimiter: CharSequence) =
-    this(delimiter.toString(), "", "")
-
-  def this(
-      delimiter: CharSequence,
-      prefix: CharSequence,
-      suffix: CharSequence
-  ) =
-    this(delimiter.toString(), prefix.toString(), suffix.toString())
+  /** Alternate constructor with no prefix or suffix */
+  def this(delimiter: CharSequence) = this(delimiter, "", "")
 
   def setEmptyValue(emptyValue: CharSequence): StringJoiner = {
     this.emptyValue = emptyValue.toString()
     this
   }
 
+  /** Utility function used in build and length. Calculates the length of
+   *  strings in the list, plus the delimiters in between them.
+   */
+  private def lengthOf(
+      parts: scala.collection.immutable.List[CharSequence],
+      delim: CharSequence
+  ): Int = {
+    val delimLength = delim.length
+    parts match {
+      case Nil => 0
+      case head :: tail =>
+        tail.foldLeft(head.length)((acc, part) =>
+          acc + delimLength + part.length
+        )
+    }
+  }
```

Comment:
```suggestion
// at the top
import java.{util => ju}
import ScalaOps._

  private def lengthOf(
      parts: ju.List[CharSequence],
      delim: CharSequence
  ): Int = {
    val delimLength = delim.length
    if (parts.length = 0) 0
    else {
      val headLen = parts(0).length
      parts.subList(1, parts.length).ScalaOps.foldLeft(headLen)((acc, part) =>
        acc + delimLength + part.length
      )
    }
  }
```
This should be close to what you need. Hope this helps.